Conch/lib/Conch/Validation.pm

```diff
@@ -0,0 +1,625 @@
+=pod
+
+=head1 NAME
+
+Conch::Validation - base class for writing Conch Validations
+
+=head1 SYNOPSIS
+
+	package Conch::Validation::DeviceValidation;
+	use Mojo::Base 'Conch::Validation';
+
+	has name        => 'device_validation';
+	has version     => 1;
+	has category    => 'CPU';
+	has description => q/Description of the validation/;
+
+	# Optional schema to validate $input_data before `validate` is run.
+	# Specified in the JSON-schema format.
+	has schema => sub {
+		{
+			required => [ 'hello' ],
+			properties => {
+				hello => { type => 'string' }
+			}
+		}
+	};
+
+	sub validate {
+		my ($self, $input_data) = @_;
+
+		my $device           = $self->device;
+		my $device_settings  = $self->device_settings;
+		my $device_location  = $self->device_location;
+		my $hardware_vendor  = $self->hardware_product_vendor;
+		my $hardware_name    = $self->hardware_product_name;
+		my $hardware_profile = $self->hardware_product_profile;
+
+		$self->register_result( expected => 'hello', got => $input_data->{hello} );
+	}
+
+
+=head1 DESCRIPTION
+
+C<Conch::Validaiton> provides the base class to define and execute Conch
+Validations. Validations extend this class by implementing a C<validate>
+method.  This method receives the input data (a C<HASHREF>) to be validatated.
+This input data hash may be validated by setting the C<schema> attribute with a
+schema definition in the L<JSON-schema|http://json-schema.org> format (Note: A
+root-level C<'object'> type is assumed in the schema. Only top-level
+properties need to be defined).
+
+The validation logic in the C<validate> method will evaluate the input data and
+register one or more validation results with the
+L<register_result|"register_result"> method. The logic may use device, device
+settings, hardware product name, hardware product vendor, and hardware product
+profile details to dispatch conditions and evaluation.
+
+Conch Validations should also define values for the C<name>, C<version>,
+C<category>, and C<description> attributes. These attributes are used in the
+identification of the validation and validation result storage in the
+Validation System infrastructure.
+
+Testing Conch Validations should be done with
+C<Test::Conch::Validation::test_validation> with TAP-based tests. This
+functions tests that Validations define the required attributes and methods,
+and allow you to test the validaiton logic by running test cases against
+expected results.
+
+=head1 METHODS
+
+=cut
+
+package Conch::Validation;
+
+use Mojo::Base -base, -signatures;
+use Mojo::Exception;
+use JSON::Validator;
+use Try::Tiny;
+
+has 'name';
+has 'version';
+has 'description';
+has 'schema';
```

Comment:
You'll see comments to this affect below but to put my larger thoughts in one place:

A lot of the validations in this PR don't use the schema to validate their input. They have custom logic inside the validation itself.  Is the JSON Schema not capable of validating those data sets? If it is capable, we need to transition those validations to use the Schema so that we don't, from day one, provide a common procedure of bypassing the Schema system entirely. If the Schema can't handle those data sets, I think we need to re-evaluate the Schema system. Basically, we need to pick a single data validation path, particularly for these first validations that will be held up as examples for future development.